src/spss/readstat_sav_read.c

```diff
@@ -145,6 +147,212 @@ static readstat_error_t sav_parse_variable_display_parameter_record(sav_ctx_t *c
 static readstat_error_t sav_parse_machine_integer_info_record(const void *data, size_t data_len, sav_ctx_t *ctx);
 static readstat_error_t sav_parse_long_string_value_labels_record(const void *data, size_t size, size_t count, sav_ctx_t *ctx);
 static readstat_error_t sav_parse_long_string_missing_values_record(const void *data, size_t size, size_t count, sav_ctx_t *ctx);
+static readstat_error_t sav_read_multiple_response_sets(size_t data_len, sav_ctx_t *ctx);
+
+static readstat_error_t parse_mr_counted_value(const char **next_part, mr_set_t *result) {
+    readstat_error_t retval = READSTAT_OK;
+    if (result->type == 'D') {
+        result->is_dichotomy = 1;
+        const char *digit_start = (*next_part);
+        while (*(*next_part) != ' ' && *(*next_part) != '\0') {
+            (*next_part)++;
+        }
+        int internal_count = (int)strtol(digit_start, NULL, 10);
+        if (*(*next_part) != ' ') {
+            retval = READSTAT_ERROR_BAD_MR_STRING;
+            goto cleanup;
+        }
+        (*next_part)++;
+        digit_start = (*next_part);
+        for (int i = 0; i < internal_count && isdigit((unsigned char)*(*next_part)); i++) {
+            (*next_part)++;
+        }
+        result->counted_value = (int)strtol(digit_start, NULL, 10);
+        if (*(*next_part) != ' ' && *(*next_part) != '\0') {
+            retval = READSTAT_ERROR_BAD_MR_STRING;
+            goto cleanup;
+        }
+    }
+    else if (result->type == 'C') {
+        result->is_dichotomy = 0;
+        result->counted_value = -1;
+    }
+    else {
+        retval = READSTAT_ERROR_BAD_MR_STRING;
+        goto cleanup;
+    }
+cleanup:
+    return retval;
+}
+
+
+static readstat_error_t parse_mr_line(const char *line, mr_set_t *result) {
```

Comment:
I think a Ragel-based parser would make it easier to follow the logic (see other files with `.rl` in this project)

Reply:
done